### PR TITLE
test windows

### PR DESCRIPTION
**Pull request by ocefpaf**
Just a few tests...

### CODE REVIEW

**Comment by ocefpaf** [comment]
Windows is failing with:

```
C:/bld/eccodes_1554681106149/work/python3/../src\grib_api_internal.h(115):
fatal error C1083: Cannot open include file: 'strings.h': No such file or directory
```

@conda-forge/eccodes, @gillins, and @isuruf do you have an idea if that is something easy to fix?

**Comment by isuruf** [comment]
Depends on which functions from strings.h is used by the package. `strings.h` has a few extra functions than `string.h` and is not available on windows. Try changing that line to `string.h` and see what happens

**Comment by ocefpaf** [comment]
> Depends on which functions from strings.h is used by the package. `strings.h` has a few extra functions than `string.h` and is not available on windows. Try changing that line to `string.h` and see what happens

Thanks! There is a

```
#ifdef  HAVE_STRING_H
 #include <string.h>
#else
 #include <strings.h>
#endif
```

in the header file so maybe we don't need to patch... Not sure how to get that to the `string.h` instead on `strings.h` though.

**Comment by ocefpaf** [comment]
Thanks @isuruf, that got a little bit further. Now I get this odd error:

```
warning: I don't know what to do with 'runtime_library_dirs':
['C:/bld/eccodes_1554685800307/work/build-python3.7/lib']
error: don't know how to set runtime library search path for MSVC
NMAKE : fatal error U1077: 'C:\bld\eccodes_1554685800307\_h_env\python.exe'
:+1: return code '0x1'
Stop.
```

**Comment by isuruf** [comment]
You need to patch `runtime_library_dirs` from setup.py. There are multiple places. Here's one, https://github.com/ecmwf/eccodes/blob/7c4a7164cb652774e21574b7110cd0f37ee678bd/python/setup.py.in#L64

**Comment by ocefpaf** [comment]
> You need to patch `runtime_library_dirs` from setup.py. There are multiple places. Here's one, [ecmwf/eccodes:python/setup.py.in@`7c4a716`#L64](https://github.com/ecmwf/eccodes/blob/7c4a7164cb652774e21574b7110cd0f37ee678bd/python/setup.py.in#L64)

Thanks! I'll give it a try.

**Comment by ocefpaf** [comment]
I removed all references to `runtime_library_dirs` to test how to re-add them one-by-one but it seems that I still get the same error message :-/

@conda-forge/eccodes team do you have an idea of what we need to build the python wrappers on Windows?

If not do you want to skip the Windows Python part and get this merged?

**Comment by isuruf** [comment]
There's more rpath lines at https://github.com/ecmwf/eccodes/blob/17458a8b1ac209e567095e896c73a4d936e5111c/python3/CMakeLists.txt#L40

File: recipe/setup_py-win.patch [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+                library_dirs=['@CMAKE_BINARY_DIR@/lib'],
+-               runtime_library_dirs=[],
++            #    runtime_library_dirs=[],
+                libraries=['eccodes'],
+                extra_compile_args=extra_compile_args,
+                extra_objects=[])
+ 
+ shared_libs='@BUILD_SHARED_LIBS@'
+-if shared_libs == 'OFF':
++# if shared_libs == 'OFF':
```
Comment (isuruf) [inline recipe/setup_py-win.patch]: `shared_libs` will be `ON`, so there's no need to patch the lines below as they are not run. Only the line 8 needs to be patched

Comment (ocefpaf) [reply recipe/setup_py-win.patch]: Thanks! Let me try that.

Comment (ocefpaf) [reply recipe/setup_py-win.patch]: Wait. I'm lost now :smile: line 8 of this file is a comment, right?

Comment (isuruf) [reply recipe/setup_py-win.patch]: I mean a patch like below,
```diff
diff --git a/python/CMakeLists.txt b/python/CMakeLists.txt
index 33aa1f0..41a1972 100644
--- a/python/CMakeLists.txt
+++ b/python/CMakeLists.txt
@@ -37,7 +37,7 @@ if( HAVE_PYTHON )
     set( _gribapi_swig "gribapi/_gribapi_swig${CMAKE_SHARED_LIBRARY_SUFFIX}" )
     # Build the extension module for use in build tree with RPATH pointing to the build tree
     add_custom_command( OUTPUT ${_gribapi_swig}
-                        COMMAND ${PYTHON_EXECUTABLE} setup.py build_ext --inplace --rpath ${CMAKE_BINARY_DIR}/lib
+                        COMMAND ${PYTHON_EXECUTABLE} setup.py build_ext --inplace
                         DEPENDS grib_interface.h grib_interface.c swig_wrap_numpy.c setup.py.in eccodes )
     add_custom_target(eccodes_build_swig_wrapper ALL DEPENDS ${_gribapi_swig})
 
@@ -56,8 +56,8 @@ if( HAVE_PYTHON )
 
     # Build the extension module for use in install tree with RPATH pointing to install tree
     install(CODE "message(\"Building Python extension modules:
-${PYTHON_EXECUTABLE} setup.py build_ext --rpath ${CMAKE_INSTALL_PREFIX}/${INSTALL_LIB_DIR}\")
-                  execute_process(COMMAND ${PYTHON_EXECUTABLE} setup.py build_ext --rpath ${CMAKE_INSTALL_PREFIX}/${INSTALL_LIB_DIR}
+${PYTHON_EXECUTABLE} setup.py build_ext\")
+                  execute_process(COMMAND ${PYTHON_EXECUTABLE} setup.py build_ext
                                   WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR})")
 
     # Library installation directory override
diff --git a/python/setup.py.in b/python/setup.py.in
index f54d428..c773e37 100644
--- a/python/setup.py.in
+++ b/python/setup.py.in
@@ -24,7 +24,7 @@ attdict = dict(sources=['@CMAKE_CURRENT_SOURCE_DIR@/swig_wrap_numpy.c',
                              '@CMAKE_CURRENT_SOURCE_DIR@/../src',
                              numpy_include],
                library_dirs=['@CMAKE_BINARY_DIR@/lib'],
-               runtime_library_dirs=[],
+               #runtime_library_dirs=[],
                libraries=['eccodes'],
                extra_compile_args=extra_compile_args,
                extra_objects=[])
diff --git a/python3/CMakeLists.txt b/python3/CMakeLists.txt
index 33aa1f0..41a1972 100644
--- a/python3/CMakeLists.txt
+++ b/python3/CMakeLists.txt
@@ -37,7 +37,7 @@ if( HAVE_PYTHON )
     set( _gribapi_swig "gribapi/_gribapi_swig${CMAKE_SHARED_LIBRARY_SUFFIX}" )
     # Build the extension module for use in build tree with RPATH pointing to the build tree
     add_custom_command( OUTPUT ${_gribapi_swig}
-                        COMMAND ${PYTHON_EXECUTABLE} setup.py build_ext --inplace --rpath ${CMAKE_BINARY_DIR}/lib
+                        COMMAND ${PYTHON_EXECUTABLE} setup.py build_ext --inplace
                         DEPENDS grib_interface.h grib_interface.c swig_wrap_numpy.c setup.py.in eccodes )
     add_custom_target(eccodes_build_swig_wrapper ALL DEPENDS ${_gribapi_swig})
 
@@ -56,8 +56,8 @@ if( HAVE_PYTHON )
 
     # Build the extension module for use in install tree with RPATH pointing to install tree
     install(CODE "message(\"Building Python extension modules:
-${PYTHON_EXECUTABLE} setup.py build_ext --rpath ${CMAKE_INSTALL_PREFIX}/${INSTALL_LIB_DIR}\")
-                  execute_process(COMMAND ${PYTHON_EXECUTABLE} setup.py build_ext --rpath ${CMAKE_INSTALL_PREFIX}/${INSTALL_LIB_DIR}
+${PYTHON_EXECUTABLE} setup.py build_ext\")
+                  execute_process(COMMAND ${PYTHON_EXECUTABLE} setup.py build_ext
                                   WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR})")
 
     # Library installation directory override
diff --git a/python3/setup.py.in b/python3/setup.py.in
index 57dc654..34ac5c4 100644
--- a/python3/setup.py.in
+++ b/python3/setup.py.in
@@ -24,7 +24,7 @@ attdict = dict(sources=['@CMAKE_CURRENT_SOURCE_DIR@/swig_wrap_numpy.c',
                              '@CMAKE_CURRENT_SOURCE_DIR@/../src',
                              numpy_include],
                library_dirs=['@CMAKE_BINARY_DIR@/lib'],
-               runtime_library_dirs=[],
+               #runtime_library_dirs=[],
                libraries=['eccodes'],
                extra_compile_args=extra_compile_args,
                extra_objects=[])
```

**Comment by dtip** [comment]
@ocefpaf It's not immediately obvious what this PR is for - is it correct that you're wanting to effectively merge the python-eccodes feedstock (https://github.com/conda-forge/python-eccodes-feedstock) into this feedstock?

As far as I'm aware, eccodes has never been built on Windows using cmake with ENABLE_PYTHON=1. Certainly the python-eccodes feedstock is disabled for Windows. I would guess ecCodes would need a little more patching to get it working, and that you'd be better off skipping the Windows build for now. But if you can get it working that would be great :)

What's the motivation for wanting to add the Windows build now? Is there a strong demand for it?

File: recipe/setup_py-win.patch (hunk 1 above)
Comment (ocefpaf) [reply recipe/setup_py-win.patch]: Thanks! I'll give it a try. (Got side tracked and forgot about this one.)

**Comment by ocefpaf** [comment]
> @ocefpaf It's not immediately obvious what this PR is for - is it correct that you're wanting to effectively merge the python-eccodes feedstock ([conda-forge/python-eccodes-feedstock](https://github.com/conda-forge/python-eccodes-feedstock)) into this feedstock?

Yep. We have better CIs resources. Merging will reduce the maintenance and help us keep it in sync.

> As far as I'm aware, eccodes has never been built on Windows using cmake with ENABLE_PYTHON=1. Certainly the python-eccodes feedstock is disabled for Windows. I would guess ecCodes would need a little more patching to get it working, and that you'd be better off skipping the Windows build for now. But if you can get it working that would be great :)

I'll give it a second try over the weekend.

> What's the motivation for wanting to add the Windows build now? Is there a strong demand for it?

`iris` moved to `eccodes` and I would love to give Windows users the same experience they as the Linux users.

**Comment by dtip** [comment]
I believe we now have Windows support over at `python-eccodes` so this one may be resolved. It's stale either way so I'll clean it up. Please re-open if you're not done!

**Pull request closed without merging**